code/datums/log_viewer.dm

```diff
@@ -0,0 +1,223 @@
+#define ALL_LOGS list(ATTACK_LOG, DEFENSE_LOG, CONVERSION_LOG, SAY_LOG, EMOTE_LOG, MISC_LOG)
+
+/datum/log_viewer
+	var/time_from = 0
+	var/time_to = 2 HOURS // Full range
```

Comment:
It should take `O(log2(n))` time to find the latest event as no event has occurred infinitely far in the future, I believe. Shouldn't be too much trouble since `log2` is a very tolerable scaling rate.

Reply:
I build a quick escape in for that case. If the end time is longer than the current `world.time` it will just return the last entry. Seeing it defaults to 4 hours now this will happen in most cases unless the user sets the time to before the current `world.time`

Reply:
Sounds good